mars/services/task/execution/mars/resource.py

```diff
@@ -12,28 +12,89 @@
+    return evaluator_cls
+
+
+def init_default_resource_for_subtask(subtask_graph: "SubtaskGraph"):  # noqa: F821
+    for subtask in subtask_graph.iter_nodes():
+        is_gpu = any(c.op.gpu for c in subtask.chunk_graph)
+        subtask.required_resource = (
+            Resource(num_gpus=1) if is_gpu else Resource(num_cpus=1)
+        )
 
 
 class ResourceEvaluator:
     """
-    Evaluate and initialize the required resource of subtasks by different
-    configurations, e.g. fixed value by default, or some recommended values
-    through external services like an HBO service which could calculate a
-    accurate value by the running history of tasks.
+    Resource evaluator is used to estimate and set resources required by
+    subtasks. It can be an internal service or an external service. If it
+    is an internal service, we can set default of adjustable resources for
+    subtasks. If it is an external service, we should report the running
+    result of the task to the external service, so that it can accurately
+    predict the required resources of subtasks based on the historical
+    running information, we call it HBO.
+
+    Best practice
+    ----------
+    You can follow the steps below to add a new resource evaluator:
+        * Inherit `ResourceEvaluator` and implement `create`, `evaluate`
+          and `report` methods. The `create` method is to create a new
+          resource evaluator instance. The `evaluate` method is to estimate
+          and set required resources for the subtasks of a task stage. And
+          this method must be implemented. The `report` method is to report
+          the running information and result of the task. And this method
+          does not have to be implemented.
+
+        * Add default configs of the new evaluator needed in `base_config.xml`
+          or its descendant files.
+
+        * Set the `resource_evaluator` to choose a resource evaluator in
+          `base_config.xml` when running a mars job.
     """
 
-    def __init__(self, stage_processor: TaskStageProcessor):
-        self._stage_processor = stage_processor
-        self._subtask_graph = stage_processor.subtask_graph
-
-    def evaluate(self):
-        """Here we could implement different acquisitions by state processor
-        configurations.
-        """
-        for subtask in self._subtask_graph.iter_nodes():
-            is_gpu = any(c.op.gpu for c in subtask.chunk_graph)
-            subtask.required_resource = (
-                Resource(num_gpus=1) if is_gpu else Resource(num_cpus=1)
+    name = None
+
+    @classmethod
+    @abstractmethod
+    async def create(cls, config: Dict[str, Any], **kwargs) -> "ResourceEvaluator":
+        name = config.get("resource_evaluator", "default")
+        evaluator_config = config.get(name, {})
+        evaluator_cls = _name_to_resource_evaluator[name]
+        if evaluator_cls.create.__func__ is ResourceEvaluator.create.__func__:
```

Comment:
Just make `ResourceEvaluator` as a abstract class? The abstractmethod has to be implemented then.

Reply:
Good suggestion. Thanks a lot.

Reply:
If `ResourceEvaluator` is a abstract class, is this check still necessary?

Reply:
It's not necessary. I'll remove these code. Thanks for your reminder.